articles/150_roslaunch.md

```diff
@@ -0,0 +1,364 @@
+However, since there's no standard tool, the launch system cannot know how to convert ROS specific declarations into a format that can be used to invoke it.
+
+Though the user could always generate the appropriate command line arguments for the process and have the launch system treat it as a normal operating system process.
+In that case, though, it might be useful to indicate to the launch system that a process may contain nodes, but should be otherwise treated like a plain operating system process.
+This would let the launch system still monitor it for ROS specific events and input/output.
+
+As an example of a "by configuration" dynamic node process, you could have a process that takes a variable number of arguments, where each is a serial port address for a laser scanner, e.g. `/dev/ACM0 /dev/ACM1 ...`, and it would instantiate a laser scanner driver node for each address.
+It would would up to this process to figure out how to allow the user to further configure the nodes though.
+It would not be possible, in this case, for the launch system to "apply a ROS namespace" to the nodes within it, for example.
+For this reason, it's probably not an ideal way to structure nodes until a standardized way to load many nodes into a single process via configuration is defined.
+
+<div class="alert alert-warning" markdown="1">
+RFC:
+
+On the point of a standardized way of having a "dynamic nodes process by configuration", I think it might be worth pursuing this, because it will basically become a more efficient way of launch a bunch of nodes into a single process, more efficient than the "by proxy" way described below.
+
+The "by proxy" way below is essentially the pattern used by nodelets in ROS 1, but what I'm describing here is essentially like a NodeletManager executable which rather than loading nodelets via Service calls would load and run them based on command line arguments or a config file or something.
+
+The "by proxy" way is more convenient when you're running things by hand, but if the roslaunch description looks something like this (just pseudo code):
+
+```xml
+<node_container_process name="my_container_process">
+  <node package="my_package" executable="talker" />
+  <node package="my_package" executable="listener" />
+
+</node_container_process>
+```
+
+Whether or not the above results in three processes (one container and two proxy processes) or a single process, probably doesn't matter to the user or at least would be fine with the user in exchange for less overhead.
+
+It would be more efficient because you don't need to have a node in each proxy to actually make the Service call to the load that node the user wants to run, and you don't need to maintain a bond between the container and the proxy processes.
+
+So I'm interested in what others think about coming up with a standardized executable which can load and configure many nodes at once from command line arguments or a config file.
```

Comment:
The above XML could be syntactic sugar for a system similar to nodelets. However, I think that means ROS2 launch would have a dependency on the container process/package, which would have a dependency on rclpy/rclcpp. Third parties implementing different language client APIs may desire a way to be launched into a container process without making ros 2 launch depend on them. If that's supported then rclcpp/rclpy should probably use the same API.

Reply:
That's a fair point, in ROS 1 this was worked around because `<node_container_process name="my_container_process">` was more like `<node pkg="nodelet" type="nodelet_manager" name="my_container_process" ...>`, i.e. it was just another node to run. If we remove the syntatic sugar, this would just be another process and if someone wanted for e.g. a rust based container for rust plugins then they would just make a new package and a new executable.

A separate concern is what happens if you put a node which is implemented as a Python script within the "node_container_process" tag? I suppose it will try to load it and fail since it's not a plugin, but it brings up the, point should we have a way to express that a `<node>` tag refers to a "process" or a dynamically loadable thing, or both?

Reply:
> That assumes we're not expecting to mix C++ and Python nodes in a single process. Whether other languages provide similar support would be up to them.

Yes, for now, this only helps C++, but it would in principle be possible to have an executable which could combine multiple Python nodes into one interpreter, or even run C++ nodes and Python nodes in a manually instantiated Python interpreter within the same process (that doesn't necessarily imply intra-process communication cross language). But I don't plan to do that right now.

Reply:
This has been resolved by our implementation of components in a separate package `launch_ros` and also due to the introduction of the concept of launch frontend and the XML frontend specifically.